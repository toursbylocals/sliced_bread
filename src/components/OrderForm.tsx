import { faker } from '@faker-js/faker';
import type { FC } from 'react';
import { useForm } from 'react-hook-form';
import { v4 } from 'uuid';

import type { ApiOrderResponseData } from '@/api/order';

import { FormError } from './FormError';
import { LabelInput } from './LabelInput';
import { LabelSelect } from './LabelSelect';
import { ProvinceStateSelect } from './ProvinceStateSelect';

type OrderFormProps = {
  updateToken: (token: string) => void;
};

export type OrderFormInput = {
  city: string;
  country: string;
  customerName: string;
  numberOfDrinks: number;
  provinceState: string;
};

export enum OrderFormField {
  City = 'city',
  Country = 'country',
  CustomerName = 'customerName',
  NumberOfDrinks = 'numberOfDrinks',
  ProvinceState = 'provinceState',
}

export const OrderForm: FC<OrderFormProps> = ({ updateToken }) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<OrderFormInput>();
  const DEFAULT_PATTERN = /^[A-Za-z -]+$/i;
  const onSubmit = async (formData: OrderFormInput) => {
    if (!formData.customerName) {
      formData.customerName = faker.person.fullName();
    }

    if (!formData.numberOfDrinks) {
      formData.numberOfDrinks = faker.number.int({ min: 1, max: 99 });
    }

    const finalData = { ...formData, orderId: v4() };

    const response = await fetch('/api/order', {
      body: JSON.stringify(finalData),
      method: 'POST',
    });
    const responseData = (await response.json()) as ApiOrderResponseData;

    updateToken(responseData.data.token);
  };

  return (
    <>
      <h5 className="font-semibold text-gray-800 mb-4">
        Order Your Spicy Manhattans!
      </h5>

      <form aria-label="form" onSubmit={handleSubmit(onSubmit)}>
        <div className="space-y-4">
          <div className="flex gap-4">
            <LabelInput
              labelFor={OrderFormField.CustomerName}
              labelText="Name"
              {...register(OrderFormField.CustomerName, {
                pattern: DEFAULT_PATTERN,
              })}
            />
            <LabelInput
              inputType="number"
              labelFor={OrderFormField.NumberOfDrinks}
              labelText="Number of Drinks"
              {...register(OrderFormField.NumberOfDrinks)}
            />
          </div>
          <div className="flex gap-4">
            <LabelInput
              labelFor={OrderFormField.City}
              labelText="City*"
              {...register(OrderFormField.City, {
                required: true,
                pattern: DEFAULT_PATTERN,
              })}
            />
            <LabelSelect
              labelFor={OrderFormField.Country}
              labelText="Country*"
              {...register(OrderFormField.Country, {
                required: true,
              })}
            >
              <option value="">Select a Country</option>
              <option value="CA">Canada</option>
              <option value="US">United States</option>
            </LabelSelect>
            <ProvinceStateSelect
              labelFor={OrderFormField.ProvinceState}
              labelText="Province / State*"
              {...register(OrderFormField.ProvinceState, {
                required: true,
              })}
            />
          </div>
          <section aria-live="polite">
            {errors.customerName && errors.customerName.type === 'pattern' && (
              <FormError>Name is invalid.</FormError>
            )}
            {errors.city && errors.city.type === 'required' && (
              <FormError>City is required.</FormError>
            )}
            {errors.country && errors.country.type === 'required' && (
              <FormError>Country is required.</FormError>
            )}
            {errors.provinceState &&
              errors.provinceState.type === 'required' && (
                <FormError>Province / State is required.</FormError>
              )}
          </section>
          <div className="flex">
            <button
              type="submit"
              className="rounded-md bg-indigo-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
            >
              Get Your Drinks!
            </button>
          </div>
        </div>
      </form>
    </>
  );
};
