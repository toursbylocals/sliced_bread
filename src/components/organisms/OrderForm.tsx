/* eslint-disable max-lines-per-function */

'use client';

import { SubmitHandler, useForm } from 'react-hook-form';
import { GridContainer } from '../atoms/GridContainer';
import { Input } from '../atoms/Input';
import { Typography } from '../atoms/Typography';
import Link from 'next/link';
import { z } from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import Button from '../atoms/Button';
import { useState } from 'react';
import { useOrderPopupStore } from '@/stores/popup';
import { motion } from 'motion/react';

const ORDER_SCHEMA = z.object({
  name: z.string(),
  quantity: z.string().regex(/^[0-9]*$/, { message: '*Numbers only' }),
  city: z.string().min(1, { message: '*Required' }),
  state: z.string().min(1, { message: '*Required' }),
  country: z.string().min(1, { message: '*Required' }),
});

type OrderInputs = z.infer<typeof ORDER_SCHEMA>;

export default function OrderForm({
  createOrder,
}: {
  createOrder: ({
    name,
    quantity,
    city,
    state,
    country,
  }: {
    name: string;
    quantity: string;
    city: string;
    state: string;
    country: string;
  }) => Promise<{
    name: string;
    quantity: number;
    city: string;
    state: string;
    country: string;
    id: string;
    created_at: Date;
  }>;
}) {
  const [loading, setLoading] = useState(false);
  const { open } = useOrderPopupStore();

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<OrderInputs>({
    defaultValues: {
      quantity: '1',
    },
    resolver: zodResolver(ORDER_SCHEMA),
  });

  // simulate real case for calling api
  const TIME_DELAY = 1000;
  const onSubmit: SubmitHandler<OrderInputs> = async (data) => {
    setLoading(true);
    const order = await createOrder({
      name: data.name || 'No Name',
      quantity: data.quantity || '1',
      city: data.city,
      state: data.state,
      country: data.country,
    });

    // simulate real case for calling api
    setTimeout(() => {
      open({
        id: order.id,
        title: 'Order has been placed!',
        description: `Remember your order id <br/><br/><strong>${order.id}</strong>`,
      });

      reset();

      setLoading(false);
    }, TIME_DELAY);
  };

  return (
    <section className="flex h-screen w-full flex-col justify-center bg-primary-300">
      <GridContainer className="relative py-16">
        <motion.div
          className="col-span-full mb-8 text-center"
          initial={{ opacity: 0, translateY: '-10%' }}
          whileInView={{
            opacity: 1,
            translateY: 0,
          }}
          transition={{ delay: 0.3, duration: 0.3 }}
        >
          <Typography variant="heading1" color="text-secondary-100">
            Taste it now
          </Typography>
        </motion.div>

        <motion.div
          className="col-span-full flex flex-col gap-8 lg:col-span-6 lg:col-start-2 xl:col-start-4"
          initial={{ opacity: 0, translateY: '-10%' }}
          whileInView={{
            opacity: 1,
            translateY: 0,
          }}
          transition={{ delay: 0.5, duration: 0.3 }}
        >
          <Link href="/privacy">
            <Typography variant="link" color="text-primary-100">
              How do we use your information?
            </Typography>
          </Link>

          <form
            className="flex flex-col gap-8"
            onSubmit={handleSubmit(onSubmit)}
          >
            <Input
              placeholder="Your Name"
              errors={errors}
              {...register('name')}
            />

            <Input
              placeholder="Quantity"
              type="number"
              errors={errors}
              {...register('quantity')}
            />
            <Input
              placeholder="City"
              errors={errors}
              {...register('city', { required: true })}
            />
            <Input
              placeholder="State/Province"
              errors={errors}
              {...register('state', { required: true })}
            />
            <Input
              placeholder="Country"
              errors={errors}
              {...register('country', { required: true })}
            />

            <Button type="submit" loading={loading}>
              <span>Get it!</span>
            </Button>
          </form>
        </motion.div>
      </GridContainer>
    </section>
  );
}
