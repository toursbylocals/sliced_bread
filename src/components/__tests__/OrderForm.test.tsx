import '@testing-library/jest-dom';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import { OrderForm } from '../OrderForm';

// Mock the updateToken function
jest.mock('uuid', () => ({
  v4: jest.fn().mockReturnValue('mock-uuid'),
}));

const mockUpdateToken = jest.fn();

const setup = () => {
  render(<OrderForm updateToken={mockUpdateToken} />);
};

describe('OrderForm Component', () => {
  it('renders the form and all required fields', () => {
    setup();

    // Check that the form title is rendered
    expect(
      screen.getByText('Order Your Spicy Manhattans!'),
    ).toBeInTheDocument();

    // Check for the inputs
    expect(screen.getByLabelText('Name')).toBeInTheDocument();
    expect(screen.getByLabelText('Number of Drinks')).toBeInTheDocument();
    expect(screen.getByLabelText('City*')).toBeInTheDocument();
    expect(screen.getByLabelText('Country*')).toBeInTheDocument();
    expect(screen.getByLabelText('Province / State*')).toBeInTheDocument();
  });

  it('displays error message when required fields are not filled out', async () => {
    setup();

    // Submit the form without filling anything
    fireEvent.submit(screen.getByRole('form'));

    // Check for error messages
    await waitFor(() => {
      expect(screen.getByText('City is required.')).toBeInTheDocument();
      expect(screen.getByText('Country is required.')).toBeInTheDocument();
      expect(
        screen.getByText('Province / State is required.'),
      ).toBeInTheDocument();
    });
  });

  it('should submit the form and call updateToken on successful submission', async () => {
    // Mock successful form submission response
    const mockResponse = {
      data: {
        token: 'mock-token',
      },
    };
    global.fetch = jest.fn().mockResolvedValue({
      json: jest.fn().mockResolvedValue(mockResponse),
    });

    setup();

    // Fill in the form with valid data
    fireEvent.input(screen.getByLabelText('Name'), {
      target: { value: 'John Doe' },
    });
    fireEvent.input(screen.getByLabelText('Number of Drinks'), {
      target: { value: '2' },
    });
    fireEvent.input(screen.getByLabelText('City*'), {
      target: { value: 'New York' },
    });
    fireEvent.change(screen.getByLabelText('Country*'), {
      target: { value: 'US' },
    });
    fireEvent.change(screen.getByLabelText('Province / State*'), {
      target: { value: 'NY' },
    });

    // Submit the form
    fireEvent.submit(screen.getByRole('form'));

    // Wait for the mock API call to complete and assert the response
    await waitFor(() => {
      expect(mockUpdateToken).toHaveBeenCalledWith('mock-token');
    });

    // Check if the fetch function was called with the correct data
    expect(global.fetch).toHaveBeenCalledWith(
      '/api/order',
      expect.objectContaining({
        method: 'POST',
        body: '{"customerName":"John Doe","numberOfDrinks":"2","city":"New York","country":"US","provinceState":"NY","orderId":"mock-uuid"}',
      }),
    );
  });

  it('displays an error when a pattern validation fails (for customerName)', async () => {
    setup();

    // Submit with invalid customer name (non-alphabetical)
    fireEvent.input(screen.getByLabelText('Name'), {
      target: { value: 'John123' },
    });

    fireEvent.submit(screen.getByRole('form'));

    await waitFor(() => {
      // Expect error to show due to pattern mismatch
      expect(screen.getByText('Name is invalid.')).toBeInTheDocument();
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<OrderForm updateToken={mockUpdateToken} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
