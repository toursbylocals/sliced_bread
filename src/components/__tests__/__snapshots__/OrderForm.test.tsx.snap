// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm Component matches snapshot 1`] = `
<DocumentFragment>
  <h5
    class="font-semibold text-gray-800 mb-4"
  >
    Order Your Spicy Manhattans!
  </h5>
  <form
    aria-label="form"
  >
    <div
      class="space-y-4"
    >
      <div
        class="flex gap-4"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="block text-sm/6 font-medium text-gray-900"
            data-testid="LabelInputLabel-customerName"
            for="customerName"
          >
            Name
          </label>
          <div
            class="mt-2"
          >
            <div
              class="flex items-center rounded-md bg-white pl-3 outline outline-1 -outline-offset-1 outline-gray-300 focus-within:outline focus-within:outline-2 focus-within:-outline-offset-2 focus-within:outline-indigo-600"
            >
              <input
                class="block min-w-0 grow py-1.5 pl-1 pr-3 text-base text-gray-900 placeholder:text-gray-400 focus:outline focus:outline-0 sm:text-sm/6"
                data-testid="LabelInputInput-customerName"
                id="customerName"
                name="customerName"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="block text-sm/6 font-medium text-gray-900"
            data-testid="LabelInputLabel-numberOfDrinks"
            for="numberOfDrinks"
          >
            Number of Drinks
          </label>
          <div
            class="mt-2"
          >
            <div
              class="flex items-center rounded-md bg-white pl-3 outline outline-1 -outline-offset-1 outline-gray-300 focus-within:outline focus-within:outline-2 focus-within:-outline-offset-2 focus-within:outline-indigo-600"
            >
              <input
                class="block min-w-0 grow py-1.5 pl-1 pr-3 text-base text-gray-900 placeholder:text-gray-400 focus:outline focus:outline-0 sm:text-sm/6"
                data-testid="LabelInputInput-numberOfDrinks"
                id="numberOfDrinks"
                name="numberOfDrinks"
                type="number"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex gap-4"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="block text-sm/6 font-medium text-gray-900"
            data-testid="LabelInputLabel-city"
            for="city"
          >
            City*
          </label>
          <div
            class="mt-2"
          >
            <div
              class="flex items-center rounded-md bg-white pl-3 outline outline-1 -outline-offset-1 outline-gray-300 focus-within:outline focus-within:outline-2 focus-within:-outline-offset-2 focus-within:outline-indigo-600"
            >
              <input
                class="block min-w-0 grow py-1.5 pl-1 pr-3 text-base text-gray-900 placeholder:text-gray-400 focus:outline focus:outline-0 sm:text-sm/6"
                data-testid="LabelInputInput-city"
                id="city"
                name="city"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="block text-sm/6 font-medium text-gray-900"
            data-testid="LabelSelectLabel-country"
            for="country"
          >
            Country*
          </label>
          <div
            class="mt-2 grid grid-cols-1"
          >
            <select
              class="col-start-1 row-start-1 w-full appearance-none rounded-md bg-white py-1.5 pl-3 pr-8 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-indigo-600 sm:text-sm/6"
              data-testid="LabelSelectSelect-country"
              id="country"
              name="country"
            >
              <option
                value=""
              >
                Select a Country
              </option>
              <option
                value="CA"
              >
                Canada
              </option>
              <option
                value="US"
              >
                United States
              </option>
            </select>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="block text-sm/6 font-medium text-gray-900"
            data-testid="LabelSelectLabel-provinceState"
            for="provinceState"
          >
            Province / State*
          </label>
          <div
            class="mt-2 grid grid-cols-1"
          >
            <select
              class="col-start-1 row-start-1 w-full appearance-none rounded-md bg-white py-1.5 pl-3 pr-8 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-indigo-600 sm:text-sm/6"
              data-testid="LabelSelectSelect-provinceState"
              id="provinceState"
              name="provinceState"
            >
              <option
                value=""
              >
                Select a Province / State
              </option>
              <optgroup
                data-testid="ProvinceStateSelect-Provinces"
                label="Provinces"
              >
                <option
                  value="AB"
                >
                  Alberta
                </option>
                <option
                  value="BC"
                >
                  British Columbia
                </option>
                <option
                  value="MB"
                >
                  Manitoba
                </option>
                <option
                  value="NB"
                >
                  New Brunswick
                </option>
                <option
                  value="NF"
                >
                  Newfoundland
                </option>
                <option
                  value="NT"
                >
                  Northwest Territories
                </option>
                <option
                  value="NS"
                >
                  Nova Scotia
                </option>
                <option
                  value="NU"
                >
                  Nunavut
                </option>
                <option
                  value="ON"
                >
                  Ontario
                </option>
                <option
                  value="PE"
                >
                  Prince Edward Island
                </option>
                <option
                  value="QC"
                >
                  Quebec
                </option>
                <option
                  value="SK"
                >
                  Saskatchewan
                </option>
                <option
                  value="YT"
                >
                  Yukon Territory
                </option>
              </optgroup>
              <optgroup
                data-testid="ProvinceStateSelect-States"
                label="States"
              >
                <option
                  value="AK"
                >
                  Alaska
                </option>
                <option
                  value="AL"
                >
                  Alabama
                </option>
                <option
                  value="AR"
                >
                  Arkansas
                </option>
                <option
                  value="AZ"
                >
                  Arizona
                </option>
                <option
                  value="CA"
                >
                  California
                </option>
                <option
                  value="CO"
                >
                  Colorado
                </option>
                <option
                  value="CT"
                >
                  Connecticut
                </option>
                <option
                  value="DC"
                >
                  District of Columbia
                </option>
                <option
                  value="DE"
                >
                  Delaware
                </option>
                <option
                  value="FL"
                >
                  Florida
                </option>
                <option
                  value="GA"
                >
                  Georgia
                </option>
                <option
                  value="HI"
                >
                  Hawaii
                </option>
                <option
                  value="IA"
                >
                  Iowa
                </option>
                <option
                  value="ID"
                >
                  Idaho
                </option>
                <option
                  value="IL"
                >
                  Illinois
                </option>
                <option
                  value="IN"
                >
                  Indiana
                </option>
                <option
                  value="KS"
                >
                  Kansas
                </option>
                <option
                  value="KY"
                >
                  Kentucky
                </option>
                <option
                  value="LA"
                >
                  Louisiana
                </option>
                <option
                  value="MA"
                >
                  Massachusetts
                </option>
                <option
                  value="MD"
                >
                  Maryland
                </option>
                <option
                  value="ME"
                >
                  Maine
                </option>
                <option
                  value="MI"
                >
                  Michigan
                </option>
                <option
                  value="MN"
                >
                  Minnesota
                </option>
                <option
                  value="MO"
                >
                  Missouri
                </option>
                <option
                  value="MS"
                >
                  Mississippi
                </option>
                <option
                  value="MT"
                >
                  Montana
                </option>
                <option
                  value="NC"
                >
                  North Carolina
                </option>
                <option
                  value="ND"
                >
                  North Dakota
                </option>
                <option
                  value="NE"
                >
                  Nebraska
                </option>
                <option
                  value="NH"
                >
                  New Hampshire
                </option>
                <option
                  value="NJ"
                >
                  New Jersey
                </option>
                <option
                  value="NM"
                >
                  New Mexico
                </option>
                <option
                  value="NV"
                >
                  Nevada
                </option>
                <option
                  value="NY"
                >
                  New York
                </option>
                <option
                  value="OH"
                >
                  Ohio
                </option>
                <option
                  value="OK"
                >
                  Oklahoma
                </option>
                <option
                  value="OR"
                >
                  Oregon
                </option>
                <option
                  value="PA"
                >
                  Pennsylvania
                </option>
                <option
                  value="PR"
                >
                  Puerto Rico
                </option>
                <option
                  value="RI"
                >
                  Rhode Island
                </option>
                <option
                  value="SC"
                >
                  South Carolina
                </option>
                <option
                  value="SD"
                >
                  South Dakota
                </option>
                <option
                  value="TN"
                >
                  Tennessee
                </option>
                <option
                  value="TX"
                >
                  Texas
                </option>
                <option
                  value="UT"
                >
                  Utah
                </option>
                <option
                  value="VA"
                >
                  Virginia
                </option>
                <option
                  value="VT"
                >
                  Vermont
                </option>
                <option
                  value="WA"
                >
                  Washington
                </option>
                <option
                  value="WI"
                >
                  Wisconsin
                </option>
                <option
                  value="WV"
                >
                  West Virginia
                </option>
                <option
                  value="WY"
                >
                  Wyoming
                </option>
              </optgroup>
            </select>
          </div>
        </div>
      </div>
      <section
        aria-live="polite"
      />
      <div
        class="flex"
      >
        <button
          class="rounded-md bg-indigo-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
          type="submit"
        >
          Get Your Drinks!
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
